core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,77 +45,113 @@ private PageFieldsToInputParametersRewriter() {}
     public static Result rewritePageFieldsToInputParameters(RowExpression expression)
     {
         Visitor visitor = new Visitor();
-        RowExpression rewrittenProjection = expression.accept(visitor, null);
-        InputChannels inputChannels = new InputChannels(visitor.getInputChannels());
+        RowExpression rewrittenProjection = expression.accept(visitor, true);
+        InputChannels inputChannels = new InputChannels(visitor.getInputChannels(), visitor.getEagerlyLoadedChannels());
         return new Result(rewrittenProjection, inputChannels);
     }
 
     private static class Visitor
-            implements RowExpressionVisitor<RowExpression, Void>
+            implements RowExpressionVisitor<RowExpression, Boolean>
     {
         private final Map<Integer, Integer> fieldToParameter = new HashMap<>();
         private final List<Integer> inputChannels = new ArrayList<>();
+        private final Set<Integer> eagerlyLoadedChannels = new HashSet<>();
         private int nextParameter;
 
         public List<Integer> getInputChannels()
         {
             return ImmutableList.copyOf(inputChannels);
         }
 
+        public List<Integer> getEagerlyLoadedChannels()
+        {
+            return ImmutableList.copyOf(eagerlyLoadedChannels);
+        }
+
         @Override
-        public RowExpression visitInputReference(InputReferenceExpression reference, Void context)
+        public RowExpression visitInputReference(InputReferenceExpression reference, Boolean unconditionallyEvaluated)
         {
-            int parameter = getParameterForField(reference);
+            int parameter = getParameterForField(reference, unconditionallyEvaluated);
             return field(parameter, reference.getType());
         }
 
-        private Integer getParameterForField(InputReferenceExpression reference)
+        private Integer getParameterForField(InputReferenceExpression reference, boolean unconditionallyEvaluated)
         {
             return fieldToParameter.computeIfAbsent(reference.getField(), field -> {
                 inputChannels.add(field);
+                if (unconditionallyEvaluated) {
+                    eagerlyLoadedChannels.add(reference.getField());
+                }
                 return nextParameter++;
             });
         }
 
         @Override
-        public RowExpression visitCall(CallExpression call, Void context)
+        public RowExpression visitCall(CallExpression call, Boolean unconditionallyEvaluated)
         {
+            boolean containsLambdaExpression = call.getArguments().stream().anyMatch(LambdaDefinitionExpression.class::isInstance);
             return new CallExpression(
                     call.getResolvedFunction(),
                     call.getArguments().stream()
-                            .map(expression -> expression.accept(this, context))
+                            // Lambda expressions maybe use only some of their input references, e.g. transform(elements, x -> 1)
```

Comment:
why `transform(elements, x -> 1)` use only some of input references?

Reply:
I'm not sure what's missing in this example but there is another example in tests : `ZIP_WITH(array_bigint0, array_bigint1, (x, y) -> 2 * x)`

Reply:
nit: The example here is really unconditional. Some conditional one could be better, but I'm fine keeping it